extensions/resteasy-reactive/quarkus-resteasy-reactive/kotlin/src/main/kotlin/org/jboss/resteasy/reactive/server/runtime/kotlin/ApplicationCoroutineScope.kt

```diff
@@ -0,0 +1,42 @@
+package org.jboss.resteasy.reactive.server.runtime.kotlin
+
+import io.vertx.core.Context
+import kotlinx.coroutines.*
+import org.eclipse.microprofile.context.ManagedExecutor
+import javax.enterprise.context.ApplicationScoped
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * Application wide coroutine scope. Should start and die with the rest of the application, along with child coroutines
+ * (structured concurrency).
+ *
+ * This should be the main interception point to place user specific [CoroutineContext.Element] objects. Things
+ * like activity ids, logger context propagation, tracing, exception handlers, etc.
+ */
+@ApplicationScoped
+class ApplicationCoroutineScope : CoroutineScope, AutoCloseable {
+    override val coroutineContext: CoroutineContext = SupervisorJob()
+
+    override fun close() {
+        coroutineContext.cancel()
+    }
+}
+
+/**
+ * Dispatches the coroutine in a worker thread from Vertx.
+ */
+class VertxDispatcher(private val vertxContext: Context) : CoroutineDispatcher() {
```

Comment:
Here is how Vert.x solves the problem:
https://github.com/vert-x3/vertx-lang-kotlin/blob/0af356249b2f63b9abf58ebf40103981d654e173/vertx-lang-kotlin-coroutines/src/main/java/io/vertx/kotlin/coroutines/VertxCoroutine.kt#L223

Reply:
Won't the proposed implementation achieve the same result without pulling another dependency?
Are there other advantages of using this approach?

Reply:
I've thought it would be better to use a battle tested version of a `CoroutineDispatcher` implementation, but after thinking a little more about it and in the light of your previous post about `asCoroutineDispatcher` i came to the conclusion that your implementation is much better 🚀 .

Reply:
What do you think of the following optimization? 

If we are on an Vert.x event thread, then we don't need to start the whole `runOnContext` maschinery.

```kotlin
class VertxDispatcher(private val vertxContext: Context) : CoroutineDispatcher() {

    override fun dispatch(context: CoroutineContext, block: Runnable) {
        if (Vertx.currentContext() != vertxContext) {
            vertxContext.runOnContext { block.run() }
        } else {
            block.run()
        }
    }
}
```

Reply:
This would violate the rule that we [mustn't call the block immediately on this method](https://github.com/Kotlin/kotlinx.coroutines/blob/master/kotlinx-coroutines-core/common/src/CoroutineDispatcher.kt#L78).

Reply:
Oh ok. Thanks for the hint.